### PR TITLE
Chapter 12: Password Reset

### PR DESCRIPTION
**Pull request by puteriffah**

### CODE REVIEW

**Comment by puteriffah** [comment]
ready

**Comment by puteriffah** [comment]
ready

File: app/assets/stylesheets/password_resets.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the PasswordResets controller here.
```
Comment (chungpv-1008) [inline app/assets/stylesheets/password_resets.scss]: remove this file because not use it

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update]
+  before_action :valid_user, only: [:edit, :update]
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  before_action :load_user, :valid_user, only: %i(edit update)
```

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      redirect_to @user
+    else
+      flash.now[:warning] = t "forget.reset.fail"
+      render :edit
+    end
+  end
+
+  private
+
+  def get_user
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  def load_user
```

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    @user = User.find_by email: params[:email]
+    return if @user
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    return if @user.present?
```

File: app/helpers/password_resets_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
```
Comment (chungpv-1008) [inline app/helpers/password_resets_helper.rb]: remove this file because not use it

File: app/models/user.rb [hunk 6]
```diff
@@ -1,6 +1,10 @@
 class User < ApplicationRecord
   USERS_PARAMS = %i(name email password password_confirmation).freeze
+<<<<<<< HEAD
   attr_accessor :remember_token, :activation_token
+=======
+  attr_accessor :remember_token, :activation_token, :reset_token
+>>>>>>> init project
```
Comment (chungpv-1008) [inline app/models/user.rb]: please fix conflict

File: app/views/password_resets/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<% provide(:title, "Reset password") %>
+
+<h1>Reset password</h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: password_reset_path(params[:id])) do |f| %>
```
Comment (chungpv-1008) [inline app/views/password_resets/edit.html.erb]:
```suggestion
    <%= form_for @user, url: password_reset_path(params[:id]) do |f| %>
```

File: app/views/password_resets/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    <%= render "shared/error_messages" %>
+    <%= hidden_field_tag :email, @user.email %>
+    <%= f.label :password, t("forget.reset.passwordlabel") %>
+    <%= f.password_field :password, class: "form-control" %>
+    <%= f.label :password_confirmation, t("forget.reset.confirmation") %>
+    <%= f.password_field :password_confirmation, class: 'form-control' %>
```
Comment (chungpv-1008) [inline app/views/password_resets/edit.html.erb]:
```suggestion
    <%= f.password_field :password_confirmation, class: "form-control" %>
```

File: app/views/user_mailer/password_reset.text.erb [hunk 9]
```diff
@@ -1,3 +1,4 @@
-User#password_reset
-
-<%= @greeting %>, find me in app/views/user_mailer/password_reset.text.erb
+<%= t "forget.reset.message" %>
+<%= edit_password_reset_url(@user.reset_token, email: @user.email) %>
```
Comment (chungpv-1008) [inline app/views/user_mailer/password_reset.text.erb]:
```suggestion
<%= edit_password_reset_url @user.reset_token, email: @user.email %>
```

File: config/routes.rb [hunk 10]
```diff
@@ -1,11 +1,14 @@
 Rails.application.routes.draw do
+  get 'password_resets/new'
+  get 'password_resets/edit'
```
Comment (chungpv-1008) [inline config/routes.rb]: using " instead of '. Move this line to the bottom

File: db/schema.rb [hunk 11]
```diff
@@ -23,6 +23,11 @@
     t.string "activation_digest"
     t.boolean "activated", default: false
     t.datetime "activated_at"
+<<<<<<< HEAD
+=======
+    t.string "reset_digest"
+    t.datetime "reset_sent_at"
+>>>>>>> init project
```
Comment (chungpv-1008) [inline db/schema.rb]: please fix conflict for this file

**Comment by puteriffah** [comment]
read

**Comment by puteriffah** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+      render :edit
+    end
+  end
+
+  private
+
+  def load_user
+    @user = User.find_by email: params[:email]
+    return if @user.present?
+    flash[:warning] = t "user.new.absent"
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    return if @user.present?
 
    flash[:warning] = t "user.new.absent"
```

File: app/controllers/password_resets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    redirect_to root_path
+  end
+
+  def valid_user
+    return if @user && @user.authenticated?(:reset, params[:id])
+      flash[:warning] = t "user.new.absent"
+      redirect_to root_path
```
Comment (chungpv-1008) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    return if @user && @user.authenticated?(:reset, params[:id])
    
    flash[:warning] = t "user.new.absent"
    redirect_to root_path
```

File: app/models/user.rb [hunk 14]
```diff
@@ -57,6 +57,15 @@ def send_activation_email
     UserMailer.account_activation(self).deliver_now
   end
 
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_columns reset_digest: FILL_IN, reset_sent_at: FILL_IN
```
Comment (chungpv-1008) [inline app/models/user.rb]: What is 'FILL_IN'

File: app/views/password_resets/edit.html.erb [hunk 15]
```diff
@@ -0,0 +1,16 @@
+<% provide(:title, "Reset password") %>
```
Comment (chungpv-1008) [inline app/views/password_resets/edit.html.erb]: remove paratheses

File: config/routes.rb [hunk 16]
```diff
@@ -6,10 +6,13 @@
...
     resources :users
     resources :account_activations, only: :edit
+    resources :password_resets, except: %i(show index destroy)
     get "/signup", to: "users#new"
     post "/signup", to: "users#create"
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     delete "/logout", to: "sessions#destroy"
+    get "password_resets/new"
+    get "password_resets/edit"
```
Comment (chungpv-1008) [inline config/routes.rb]: remove

**Comment by puteriffah** [comment]
ready

**Comment by vanvtt-0952** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 17]
```diff
@@ -0,0 +1,53 @@
+class PasswordResetsController < ApplicationController
+  before_action :load_user, :valid_user, only: %i(edit update)
+  before_action :valid_user, :check_expiration, only: %i(edit update)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]: They have same action, therefore you can write in one line.

File: app/controllers/password_resets_controller.rb [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+    flash[:warning] = t "user.new.absent"
+    redirect_to root_path
+  end
+
+  def valid_user
+    return if @user && @user.authenticated?(:reset, params[:id])
+
+      flash[:warning] = t "user.new.absent"
+      redirect_to root_path
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    flash[:warning] = t "user.new.absent"
    redirect_to root_path
```

**Comment by puteriffah** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**